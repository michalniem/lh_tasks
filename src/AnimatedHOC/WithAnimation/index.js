import React from "react";
import { motion } from "framer-motion";

const defaultAnimationOptions = {
  enter: { opacity: 1 },
  exit: { opacity: 0 },
  transition: { duration: 0.3, delay: 0.5 },
};

const defaultIntersectionOptions = {
  threshold: [1],
  rootMargin: "-100px 0px -100px 0px",
};

function WithAnimation(Component, propsToPass) {
  return (
    animation = defaultAnimationOptions,
    intersection = defaultIntersectionOptions
  ) =>
    class extends React.Component {
      constructor(props) {
        super(props);
        this.state = {
          isIntersecting: false,
        };
        this.rootRef = React.createRef();
        this.observer = new IntersectionObserver(
          ([entry]) => this.observerCallback(entry),
          intersection
        );
      }

      observerCallback = (entry) => {
        this.setState({ isIntersecting: entry.isIntersecting });
      };

      componentDidMount() {
        if (this.rootRef.current) this.observer.observe(this.rootRef.current);
      }

      componentWillUnmount() {
        if (this.rootRef.current) this.observer.unobserve(this.rootRef.current);
      }

      render() {
        return (
          <div ref={this.rootRef}>
            <motion.div
              data-test-id="animated_element"
              animate={this.state.isIntersecting ? "enter" : "exit"}
              variants={{ enter: animation.enter, exit: animation.exit }}
              transition={animation.transition}
            >
              <Component {...propsToPass} />
            </motion.div>
          </div>
        );
      }
    };
}

export default WithAnimation;
