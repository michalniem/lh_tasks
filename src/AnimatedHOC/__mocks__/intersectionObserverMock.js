export function setupIntersectionObserverMock({
  observe = () => null,
  unobserve = () => null,
} = {}) {
  class IntersectionObserver {
    observe = observe;
    unobserve = unobserve;
  }
  Object.defineProperty(
    global,
    'IntersectionObserver',
    { writable: true, configurable: true, value: IntersectionObserver }
  );
}