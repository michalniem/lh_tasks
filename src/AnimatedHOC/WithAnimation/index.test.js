import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

import WithAnimation from "./index";

import { setupIntersectionObserverMock } from "../__mocks__/intersectionObserverMock";

const NoopComponent = () => <div />;

describe("WithProgressBar", () => {
  beforeEach(() => {
    setupIntersectionObserverMock();
  });

  test("should change isIntersecting state to true", () => {
    const Component = WithAnimation(NoopComponent)();
    const wrapper = shallow(<Component />);

    const instance = wrapper.instance();
    instance.observerCallback({ isIntersecting: true });

    expect(wrapper.state("isIntersecting")).toBeTruthy();
  });

  test("should change isIntersecting state after observerCallback invoke", () => {
    const Component = WithAnimation(NoopComponent)();
    const wrapper = shallow(<Component />);

    const instance = wrapper.instance();
    instance.observerCallback({ isIntersecting: true });

    expect(wrapper.state("isIntersecting")).toBeTruthy();
  });

  test("should set animate variant to enter if isIntersecting state is truthy", () => {
    const Component = WithAnimation(NoopComponent)();
    const wrapper = shallow(<Component />);
    wrapper.setState({ isIntersecting: true });
    const animationVariant = wrapper
      .find("[data-test-id='animated_element']")
      .prop("animate")

    expect(animationVariant).toEqual("enter");
  });

  test("should set animate variant to exit if isIntersecting state is falsy", () => {
    const Component = WithAnimation(NoopComponent)();
    const wrapper = shallow(<Component />);
    wrapper.setState({ isIntersecting: false });
    const animationVariant = wrapper
      .find("[data-test-id='animated_element']")
      .prop("animate")

    expect(animationVariant).toEqual("exit");
  });

  test("should render correctly", () => {
    const Component = WithAnimation(NoopComponent)();
    const wrapper = shallow(<Component />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
