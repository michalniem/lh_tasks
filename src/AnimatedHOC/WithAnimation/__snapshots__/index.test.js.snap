// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithProgressBar should render correctly 1`] = `
<div>
  <ForwardRef(MotionComponent)
    animate="exit"
    data-test-id="animated_element"
    transition={
      Object {
        "delay": 0.5,
        "duration": 0.3,
      }
    }
    variants={
      Object {
        "enter": Object {
          "opacity": 1,
        },
        "exit": Object {
          "opacity": 0,
        },
      }
    }
  >
    <NoopComponent />
  </ForwardRef(MotionComponent)>
</div>
`;
